src/otr/module.c

```diff
@@ -0,0 +1,269 @@
+ */
+
+#define _GNU_SOURCE
+#include <glib.h>
+
+#include "module.h"
+
+#include "signals.h"
+#include "queries.h"
+#include "commands.h"
+
+#include "irc.h"
+#include "irc-servers.h"
+#include "irc-queries.h"
+#include "irc-commands.h"
+
+#include "key.h"
+#include "otr.h"
+#include "otr-formats.h"
+#include "otr-fe.h"
+#include "misc.h"
+
+int debug = FALSE;
+
+/*
+ * Global state for the user. Init when the module loads.
+ */
+struct otr_user_state *user_state_global;
+
+/*
+ * Pipes all outgoing private messages through OTR
+ */
+static void sig_server_sendmsg(SERVER_REC *server, const char *target,
+		const char *msg, void *target_type_p)
+{
+	char *otrmsg = NULL;
+
+	if (GPOINTER_TO_INT(target_type_p) != SEND_TARGET_NICK) {
+		otrl_message_free(otrmsg);
+		return;
+	}
+
+	/* Critical section. On error, message MUST NOT be sent */
+	if (otr_send(server, msg, target, &otrmsg)) {
+		signal_stop();
+		otrl_message_free(otrmsg);
+		return;
+	}
+
+	if (otrmsg == NULL) {
+		/* Send original message */
+		signal_continue(4, server, target, msg, target_type_p);
+	} else {
+		/* Send encrypted message */
+		signal_continue(4, server, target, otrmsg, target_type_p);
+	}
+
+	otrl_message_free(otrmsg);
+}
+
+/*
+ * Pipes all incoming private messages through OTR
+ */
+void sig_message_private(SERVER_REC *server, const char *msg, const char *nick, const char *address)
+{
+	char *new_msg = NULL;
+
+	if (otr_receive(server, msg, nick, &new_msg)) {
+		signal_stop();
+		otrl_message_free(new_msg);
+		return;
+	}
+
+	if (new_msg == NULL) {
+		/* This message was not OTR */
+		signal_continue(4, server, msg, nick, address);
+	} else {
+		/*
+		 * Check for /me IRC marker and if so, handle it so the user does not
+		 * receive a message beginning with /me but rather let irssi handle it
+		 * as a IRC action.
+		 */
+		if (strncmp(new_msg, OTR_IRC_MARKER_ME, OTR_IRC_MARKER_ME_LEN) == 0) {
+			signal_stop();
+			signal_emit("message irc action", 5, server, new_msg + OTR_IRC_MARKER_ME_LEN, nick, address, nick);
+		} else {
+			/* OTR received message */
+			signal_continue(4, server, new_msg, nick, address);
+		}
+	}
+
+	otrl_message_free(new_msg);
+}
+
+/*
+ * Finish an OTR conversation when its query is closed.
+ */
+static void sig_query_destroyed(QUERY_REC *query)
+{
+	if (query && query->server && query->server->connrec) {
+		otr_finish(query->server, query->name);
+	}
+}
+
+/*
+ * Handle /me IRC command.
+ */
+static void cmd_me(const char *data, IRC_SERVER_REC *server,
+		WI_ITEM_REC *item)
+{
+	int ret;
+	const char *target;
+	char *msg, *otrmsg = NULL;
+	QUERY_REC *query;
+
+	query = QUERY(item);
+
+	if (query == NULL || query->server == NULL) {
+		return;
+	}
+
+	CMD_IRC_SERVER(server);
+	if (!IS_IRC_QUERY(query)) {
+		return;
+	}
+
+	if (server == NULL || !server->connected) {
+		cmd_return_error(CMDERR_NOT_CONNECTED);
+	}
+
+	target = window_item_get_target(item);
+
+	ret = asprintf(&msg, OTR_IRC_MARKER_ME "%s", data);
```

Comment:
I'd replace every asprintf with `g_strdup_printf()`, more portable

Reply:
Fixed in c43af47f2c8124210bc0ced8ef05555b18ccccab